Address issues with unexpected MODULE values

ktbarrett opened the pull request:
Fixes #1591. Also addresses issue with `HOOKS`.

ktbarrett commented:
Added support for trailing and leading whitespace in elements.

imphil commented:
Code looks good modulo the empty MODULE error message change. Let's discuss that in https://github.com/cocotb/cocotb/issues/1591 and then return to this PR.

eric-wieser commented:
I put some comments on  #1591.

eric-wieser commented:
A concern orthogonal to the discussion there - I'm not convinced naming tests by their issue numbers are useful, unless the corner case is hard to describe concisely. More useful test names would be:

* `test_module_var_unset`
* `test_module_var_empty`
* `test_module_without_tests`

ktbarrett commented:
Addressed comments here and in the linked issue.

If module is not set, we throw an error and exist uncleanly (without result.xml).
If module is empty (or no tests were found in any of the modules), we warn that no tests were run and exit cleanly (with result.xml).

Because of https://github.com/cocotb/cocotb/issues/1591#issuecomment-610414495, I removed the test for unset `MODULE` because the makefiles make testing that impossible right now.

ktbarrett commented:
@imphil We can start returning non-0 exit codes once we move to something like #1562. The regression system will have control start to finish and the simulator will sit in the middle.

imphil commented:
Yeah, that's a whole different topic.

eric-wieser commented on cocotb/__init__.py:
```diff
@@ -164,12 +164,12 @@ def _initialise_testbench(root_name):
 
-    if not module_str:
+    if module_str is None:
         raise ImportError("Environment variables defining the module(s) to " +
                           "execute not defined.  MODULE=\"%s\"" % (module_str))
```
This is dumb, if the module string was not defined then there's no poitn printing it, especially if we're going to imply it was set to the string `None`.
```suggestion
        raise ImportError("Environment variables MODULE which defines the module(s) to 
                          "execute is not defined.")
```

ktbarrett replied:
I feel like `ImportError` is also improper, will change to `ValueError`.

eric-wieser replied:
Hadn't even noticed that, good idea.

eric-wieser commented on cocotb/regression.py:
```diff
@@ -180,6 +181,10 @@ def initialise(self):
                     self._init_test(thing)
+                    have_tests = True
+
+        if not have_tests:
```
Could you use `if self._queue` here? Or does that not work?

ktbarrett replied:
Thing are only added to the queue if they initialize properly and are not skipped. What we want to see is if we find anything marked as a `test`.

eric-wieser replied:
I'd perhaps argue we want a warning for either case, but I'm happy to call that out of scope.

eric-wieser commented on tests/test_cases/test_module_var_empty/Makefile:
```diff
@@ -0,0 +1,6 @@
+# test MODULE is empty
+
+MODULE := ""
+
```
nit:
```suggestion
```

eric-wieser commented on tests/test_cases/test_module_without_tests/Makefile:
```diff
@@ -0,0 +1,5 @@
+include ../../designs/sample_module/Makefile
+
+MODULE=test_nothing
+
+# test MODULE is set, but there are no tests across any of the modules
```
nit, to match the others:
```suggestion
# test MODULE is set, but there are no tests across any of the modules

MODULE=test_nothing
```

eric-wieser commented on cocotb/__init__.py:
```diff
@@ -164,12 +164,11 @@ def _initialise_testbench(root_name):
-        raise ImportError("Environment variables defining the module(s) to " +
-                          "execute not defined.  MODULE=\"%s\"" % (module_str))
+    if module_str is None:
+        raise ValueError("Environment variable MODULE, which defines the module(s) to execute, is not defined.")
```
Line's a little long, but whatever

ktbarrett commented:
So this is weird.

https://travis-ci.org/github/cocotb/cocotb/jobs/672124226#L4055-L4083

`MODULE` is set to the empty string, but the "MODULE not set" error is raised, indicating that `getenv` returned `None` instead of the empty string. Does `getenv` or the environment work differently on Windows? I don't know much of anything about Windows.

eric-wieser commented:
Sounds like you ran into https://github.com/msys2/MINGW-packages/issues/1599

eric-wieser commented:
CI failure is #1603

eric-wieser commented:
Does this justify a newsfragment? In that whitespace is now stripped in the `MODULE` variable, meaning it can be wrapped across multiple lines?

ktbarrett commented:
@eric-wieser I'll do some additional cleanup on other env var parsing in a later PR and I'll make a note then saying they are all whitespace and empty element tolerant.
